### DISCUSSION TITLE
Bare metal deployment mechanics

### DISCUSSION THREAD

**Discussion by jmpolom** [systemd/mkosi#2565] [General]
If you use mkosi today, what are the options like for deploying the resulting image(s) to a bare metal system? How do these options change if you want to LUKS encrypt the system volume(s)? Seems like this would all be possible but there are options so I'm curious what people have tried.

**Comment by daandemeyer** [comment]
@jmpolom Deploying to bare metal is hampered by lack of an end-to-end update solution. systemd-sysupdate will fill this gap but I haven't had the chance to figure out smooth integration of it with mkosi yet.

It works great for rescue or installer images already though. Use `mkosi burn` to install the image to a USB, make sure linux-firmware is installed in the image, add `KernelCommandLine=` to get rid of the default `console=ttyS0` and it should boot.

**Reply by jmpolom** [reply]
I was figuring I would investigate prospects with sysupdate so that was not my issue. My main question is for the initial deployment (blank storage media), what are some methods for getting the resultant image onto hardware based storage in a system (conventional x86 EFI no embedded stuff)? Just dd it to the target block device from a live booted environment and rely on repart to setup partitions properly on first boot?

I realize this is a bit of a brainstorming question. Looking for ideas not finished solutions.

**Reply by daandemeyer** [reply]
Instead of `dd`, you would do `systemd-repart --copy-from /dev/sdb /dev/sda` which will handle sector size a little better.

But repart will still try to run on your USB stick if put repart partition definitions in the image. We're going to add some kind of installer mode to systemd to deal with this but this is still very WIP and not merged upstream.

**Reply by jmpolom** [reply]
Can repart actions be tied to firstboot or disabled via kargs?

**Reply by daandemeyer** [reply]
@jmpolom Nope. Can't tie it to firstboot as that means it won't run on your deployed image. `systemd-repart --copy-from /dev/sdb /dev/sda` would only copy the usr and ESP partitions. You would still need repart to run on first boot after installing to create the TPM encrypted root partition. Can't disable via kargs because you're supposed to use secure boot which means you can't modify the kernel command line.

**Reply by jmpolom** [reply]
Is UKI the only kernel format supported with mkosi? Is a conventional /boot with individual kernel, initrd and grub not supported?

**Reply by daandemeyer** [reply]
> Is UKI the only kernel format supported with mkosi? Is a conventional /boot with individual kernel, initrd and grub not supported?

It is supported, see UnifiedKernelImages setting. But that doesn't matter for what I said above, since any solution we come up with needs to work for locked down systems as well.

**Reply by jmpolom** [reply]
Somehow I missed that option in the man page.

> `systemd-repart --copy-from /dev/sdb /dev/sda` would only copy the usr and ESP partitions.

Why would repart only copy the /usr and ESP partitions with the command example you provided? Based on reading the repart man page with the `--copy-from` option this is far from an intuitive/obvious outcome.

**Reply by daandemeyer** [reply]
I'm assuming those are the only partitions on the USB and that the others are created on first boot

**Reply by jmpolom** [reply]
Does mkosi only create a /usr and ESP partition by default? The documentation suggests you'll get a full directory tree unless you configure it otherwise?

**Reply by daandemeyer** [reply]
Yes you get full control via mkosi.repart/

**Comment by jmpolom** [comment]
> systemd-repart --copy-from /dev/sdb /dev/sda

The man page describes an [IMAGE argument](https://www.freedesktop.org/software/systemd/man/latest/systemd-repart.html#--copy-from=IMAGE) for the --copy-from option. Should this work to get repart to copy a disk image to a block device? The supported image formats aren't described in the man page but I'm assuming a raw image works?
